external-crates/move/crates/move-compiler/src/typing/deprecations.rs

```diff
@@ -0,0 +1,765 @@
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    diag,
+    diagnostics::codes::TypeSafety,
+    expansion::ast::{self as E, ModuleIdent},
+    naming::ast::{self as N, StructFields, Type, Type_},
+    parser::ast::{ConstantName, DatatypeName, FunctionName},
+    shared::{
+        known_attributes::{AttributePosition, KnownAttribute},
+        CompilationEnv, Identifier, Name,
+    },
+    typing::{ast as T, visitor::TypingVisitorContext},
+};
+use move_ir_types::location::*;
+use move_proc_macros::growing_stack;
+use std::collections::{BTreeMap, BTreeSet};
+
+const NOTE_STR: &str = "note";
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct Deprecation {
+    // The source location of the deprecation attribute
+    source_location: Loc,
+    // The type of the member that is deprecated (function, constant, etc.)
+    location: AttributePosition,
+    // The module that the deprecated member belongs to. This is used in part to make sure we don't
+    // register deprecation warnings for members within a given deprecated module calling within
+    // that module.
+    module_ident: ModuleIdent,
+    // Information about the deprecation information depending on the deprecation attribute.
+    deprecation_info: DeprecationInfo,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+enum DeprecationInfo {
+    // #[deprecated]
+    Deprecated,
+    // #[deprecated(note = b"message")]
+    DeprecatedWithNote(String),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct DeprecationId {
+    mident: ModuleIdent,
+    name: Name,
+    deprecation: Deprecation,
```

Comment:
This is because I'm keying "warning locations" in the map by the pair of `(named_context, deprecation)` where `named_context = mident::name`. 

Basically for each function/struct/const (etc) and deprecation pair we group all the warnings to emit using this key.